status/status_ext_test.go

```diff
@@ -80,3 +90,161 @@ func (s) TestErrorIs(t *testing.T) {
+					grpc.SetTrailer(ctx, <-trailerCh)
+					return nil, <-errCh
+				},
+			}
+			if err := serverType.startServerFunc(ss); err != nil {
+				t.Fatalf("Error starting endpoint server: %v", err)
+			}
+			if err := ss.StartClient(); err != nil {
+				t.Fatalf("Error starting endpoint client: %v", err)
+			}
+			defer ss.Stop()
+
+			// Convenience function for making a status including details.
+			detailErr := func(c codes.Code, m string) error {
+				s, err := status.New(c, m).WithDetails(&testpb.SimpleRequest{
+					Payload: &testpb.Payload{Body: []byte("detail msg")},
+				})
+				if err != nil {
+					t.Fatalf("Error adding details: %v", err)
+				}
+				return s.Err()
+			}
+
+			serialize := func(err error) string {
+				buf, _ := proto.Marshal(status.Convert(err).Proto())
+				return string(buf)
+			}
+
+			testCases := []struct {
+				name        string
+				trailerSent metadata.MD
+				errSent     error
+				trailerWant []string
+				errWant     error
+				errContains error
+			}{{
+				name:        "basic without details",
+				trailerSent: metadata.MD{},
+				errSent:     status.Error(codes.Aborted, "test msg"),
+				errWant:     status.Error(codes.Aborted, "test msg"),
+			}, {
+				name:        "basic without details passes through trailers",
+				trailerSent: metadata.MD{"grpc-status-details-bin": []string{"random text"}},
+				errSent:     status.Error(codes.Aborted, "test msg"),
+				trailerWant: []string{"random text"},
+				errWant:     status.Error(codes.Aborted, "test msg"),
+			}, {
+				name:        "basic without details conflicts with manual details",
+				trailerSent: metadata.MD{"grpc-status-details-bin": []string{serialize(status.Error(codes.Canceled, "test msg"))}},
+				errSent:     status.Error(codes.Aborted, "test msg"),
+				trailerWant: []string{serialize(status.Error(codes.Canceled, "test msg"))},
+				errContains: status.Error(codes.Internal, "mismatch"),
+			}, {
+				name:        "basic with details",
+				trailerSent: metadata.MD{},
+				errSent:     detailErr(codes.Aborted, "test msg"),
+				trailerWant: []string{serialize(detailErr(codes.Aborted, "test msg"))},
+				errWant:     detailErr(codes.Aborted, "test msg"),
+			}, {
+				name:        "basic with details discards user's trailers",
+				trailerSent: metadata.MD{"grpc-status-details-bin": []string{"will be ignored"}},
+				errSent:     detailErr(codes.Aborted, "test msg"),
+				trailerWant: []string{serialize(detailErr(codes.Aborted, "test msg"))},
+				errWant:     detailErr(codes.Aborted, "test msg"),
+			}}
+
+			for _, tc := range testCases {
+				t.Run(tc.name, func(t *testing.T) {
+					trailerCh <- tc.trailerSent
+					errCh <- tc.errSent
+					trailerGot := metadata.MD{}
+					_, errGot := ss.Client.UnaryCall(ctx, &testpb.SimpleRequest{}, grpc.Trailer(&trailerGot))
+					gsdb := trailerGot["grpc-status-details-bin"]
+					if !reflect.DeepEqual(gsdb, tc.trailerWant) {
```

Comment:
Any reason for this to not use `cmp.Equal`?

Reply:
> why prefer `cmp.Equal` over `reflect.DeepEqual`?

Because it is the recommended way of doing comparisons in tests.

> if there are no special equality options needed?

We shouldn't have to think about whether special comparison options are required before making the decision between the two, every time we have to make a comparison in our tests. `reflect.DeepEqual` is sensitive to changes in unexported fields. Agreed that we are only comparing a slice of strings here, but I feel more usages of `reflect.DeepEqual` will lead to more of them in our codebase (saying I'm using it to maintain consistency, or oh that test does it and why shouldn't this one do?)

Reply:
> Because it is the recommended way of doing comparisons in tests.

The rationale cites "reflect.DeepEqual compares unexported fields <which presumably shouldn't ever be important to your tests I guess?>"...but if unexported fields is what you want isn't `reflect.DeepEqual` much easier to use?  In this case it's comparing `[]string` so if it isn't good enough then I don't understand why.

If `cmp` is preferred why isn't it in the stdlib?  🤷 

I switched it but don't really see the rationale of always reaching for something that's often more complicated and harder to use by default over something that for simple use cases is easier to use and understand.

Reply:
I see that commit where you switched, but for some reason github doesn't show it in the full diff.